igvc_navigation/src/trajectory_planner/motion_profiler.cpp

```diff
@@ -0,0 +1,86 @@
+#include "motion_profiler.h"
+#include <tf/transform_datatypes.h>
+
+MotionProfiler::MotionProfiler(double axle_length, const WheelConstraint& wheel_constraint,
+                               const RobotConstraint& robot_constraint,
+                               const MotionProfilerOptions& motion_profiler_options, double target_velocity)
+  : wheel_constraint_{ wheel_constraint }
+  , robot_constraint_{ robot_constraint }
+  , motion_profiler_options_{ motion_profiler_options }
+  , axle_length_{ axle_length }
+  , target_velocity_{ target_velocity }
+{
+}
+
+void MotionProfiler::profileTrajectory(const igvc_msgs::trajectoryPtr& trajectory_ptr, const RobotState& state)
+{
+  // For now, a simple "motion profile" where velocity is a function of curvature. ie. as curvature increases,
+  // velocity decreases.
+  trajectory_ptr->trajectory.front().header.stamp = trajectory_ptr->header.stamp;
+
+  for (size_t i = 0; i < trajectory_ptr->trajectory.size(); i++)
+  {
+    igvc_msgs::trajectory_point& trajectory_point = trajectory_ptr->trajectory[i];
+    double beta = motion_profiler_options_.beta;
+    double lambda = motion_profiler_options_.lambda;
+    double velocity = target_velocity_ / (1 + beta * std::pow(std::abs(trajectory_point.curvature), lambda));
+
+    // Constraint checking and limiting
+    if (velocity > robot_constraint_.velocity)
+    {
+      velocity = robot_constraint_.velocity;
+    }
+
+    if (i == 0)
+    {
+      trajectory_point.velocity = state.linearVelocity();
+    }
+    else
+    {
+      trajectory_point.velocity = velocity;
+    }
+
+    // Caclulate time to next point
+    if (i < trajectory_ptr->trajectory.size() - 1)
+    {
+      double distance;
+      if (std::abs(trajectory_point.curvature) > 1e-8)
+      {
+        double R = 1 / trajectory_point.curvature;
+        double current_yaw = tf::getYaw(trajectory_point.pose.orientation);
+        double next_yaw = tf::getYaw(trajectory_ptr->trajectory[i + 1].pose.orientation);
+
+        double d_theta = next_yaw - current_yaw;
+        distance = std::abs(d_theta * R);
```

Comment:
This calculation should be very close to the below case for small delta yaw. Does it really need to be a separate case? (Hint: curvature was _probably_ calculated using the (dyaw)/sqrt(dx^2 + dy^2) formulation, so this really doesn't change anything).

If the answer to the above is yes, please actually mention what it does in a comment - i.e. "Approximate this segment of the path as an arc with radius `1 / curvature`".

Reply:
yeah, this makes sense. Even though it doesn't _need_ to be a separate case, I think that we don't really lose anything by adding it in. I'll add a comment to clarify things